cpu/atmega_common/periph/rtt.c

```diff
@@ -39,28 +54,99 @@
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
+#ifdef RTT_BACKEND_SC
+/**
+ * Atomic read of 32-bit registers according
+ * to section 10.3 of the Datasheet
+ * @{
+ */
+
+__attribute__((always_inline))
+  static inline uint32_t rg_read32(volatile uint8_t *hh,
+                                   volatile uint8_t *hl,
+                                   volatile uint8_t *lh,
+                                   volatile uint8_t *ll)
+{
+    le_uint32_t rg;
+
+    rg.u8[0] = *ll;
+    rg.u8[1] = *lh;
+    rg.u8[2] = *hl;
+    rg.u8[3] = *hh;
+
+    return rg.u32;
+}
```

Comment:
The `__attribute__((always_inline))` should not be needed, unless someone compiles with `-O0`. But anyone doing so will expect additional overhead in CPU cycles and ROM size anyway...

This could be more simplified, as the layout of those 32 bit registers are always consecutive with the least significant byte having the lowest address.

The current implementation contains a data race: According to the data sheet a single 32 bit temporary register is used to capture the current contents of all available 32 bit registers. And a read of the least significant byte of any 32 bit registers cause the temporary register to be updated. Let's say thread A starts to read a 32 bit registers and has read two bytes, than an IRQ is triggered. If the ISR is now reading a 32 bit registers, it will cause the temporary register to be updated. After the ISR completes, thread A will continue to read the remaining two bytes, but those will return the updated contents of the temporary register.

Also, a short summary of the reason the accesses are atomic as a comment would make the life of people reading the code easier. E.g:

```C
/*
 * Read a 32 bit register as described in section 10.3 of the datasheet: A read
 * of the least significant byte causes the current value to be atomically
 * captured in atemporary 32 bit registers. The remaining reads will access this
 * register instead. Only a single 32 bit temporary register is used to provide 
 * means to atomically access them. Thus, interrupts must be disabled during the
 * read sequence in order to prevent other threads (or ISRs) from updating the
 * temporary 32 bit register before the reading sequence has completed.
 */
static inline uint32_t reg32_read(volatile uint8_t *reg_ll)
{
    le_uint32_t reg;
    unsigned state = irq_disable();
    reg.u8[0] =  reg_ll[0];
    reg.u8[1] =  reg_ll[1];
    reg.u8[2] =  reg_ll[2];
    reg.u8[3] =  reg_ll[3];
    irq_restore(state);
    return reg.u32;
}
```

Reply:
Thank you.

Reply:
The implementation only calls `rg_read32()` when interrupts are already disabled though, no need to disable them again.

Reply:
Wait, in `rtt_get_alarm()` it does so. But that is the only exception; so adding `irq_disable()` and `irq_restore()` there and dropping it here would indeed be better.

Reply:
It's a pity that we cannot tell the compiler the semantics of `irq_disable()` and `irq_restore()` to allow it to just optimize out all inner pairs of `irq_disable()` and `irq_restore()` when inlining a function.

Reply:
But the alarm value is not being modified by the MCU.
Only reading `SCCNT` is racy because this is continuously changing.

Reply:
Seems these functions should be extracted to another perepherial module named "sc"  as they could be used by optimized 802.15.4 network drivers to read other 32-bit registers related to the MAC symbol counter.

Reply:
> But the alarm value is not being modified by the MCU.

Doesn't matter. It is read throw the 32 bit temporary register. And if while reading the temporary registers another thread or ISR is executed, the contents of the temporary register might be modified.